okhttp/src/main/java/okhttp3/EventListener.java

```diff
@@ -61,7 +63,18 @@ public void dnsEnd(Call call, String domainName, @Nullable List<InetAddress> ine
       @Nullable Throwable throwable) {
   }
 
-  public void connectStart(Call call, InetAddress address, int port) {
+  /**
+   * Invoked just prior to initiating a socket connection.
+   *
+   * <p>This method will be invoked if no existing connection in the {@link ConnectionPool} can be
+   * reused.
+   *
+   * <p>This can be invoked more than 1 time for a single {@link Call}. For example, if the response
+   * to the {@link Call#request()} is a redirect to a different address, or a connection is retried.
+   *
+   * <p>{@code proxy} will be non-null if an HTTP or SOCKS proxy is used and null otherwise.
```

Comment:
(elsewhere we use `null` to indicate “we haven’t selected a proxy” and `NO_PROXY` to indicate “we’ve decided to use no proxy”)